CMakeLists.txt

```diff
@@ -1,187 +1,233 @@
-# FREERTOS_CONFIG_FILE_DIRECTORY
-# FREERTOS_PORT
+# User is responsible to one mandatory option:
+#   FREERTOS_PORT, if not specified and native port detected, uses the native compile.
+#
+# User is responsible for one library target:
+#   freertos_config ,typcially an INTERFACE library
 #
 # User can choose which heap implementation to use (either the implementations
 # included with FreeRTOS [1..5] or a custom implementation ) by providing the
 # option FREERTOS_HEAP. If the option is not set, the cmake will default to
 # using heap_4.c.
 
-# Absolute path to FreeRTOS config file directory
-set(FREERTOS_CONFIG_FILE_DIRECTORY "" CACHE STRING "Absolute path to the directory with FreeRTOSConfig.h")
+# `freertos_config` target defines the path to FreeRTOSConfig.h and optionally other freertos based config files
+if(NOT TARGET freertos_config )
+    if (NOT DEFINED FREERTOS_CONFIG_FILE_DIRECTORY )
 
-if(NOT FREERTOS_CONFIG_FILE_DIRECTORY)
-    message(FATAL_ERROR " FreeRTOSConfig.h file directory not specified. Please specify absolute path to it from top-level CMake file:\n"
-        "  set(FREERTOS_CONFIG_FILE_DIRECTORY <absolute path to FreeRTOSConfig.h directory> CACHE STRING \"\")\n"
-        " or from CMake command line option:\n"
-        "  -DFREERTOS_CONFIG_FILE_DIRECTORY='/absolute_path/to/FreeRTOSConfig.h/directory'")
-elseif(NOT EXISTS ${FREERTOS_CONFIG_FILE_DIRECTORY}/FreeRTOSConfig.h)
-    message(FATAL_ERROR " FreeRTOSConfig.h file not found in the directory specified (${FREERTOS_CONFIG_FILE_DIRECTORY})\n"
-        " Please specify absolute path to it from top-level CMake file:\n"
-        "  set(FREERTOS_CONFIG_FILE_DIRECTORY <absolute path to FreeRTOSConfig.h directory> CACHE STRING \"\")\n"
-        " or from CMake command line option:\n"
-        "  -DFREERTOS_CONFIG_FILE_DIRECTORY='/absolute_path/to/FreeRTOSConfig.h/directory'")
+        message(FATAL_ERROR " freertos_config target not specified.  Please specify a cmake target that defines the include directory for FreeRTOSConfig.h:\n"
+            "  add_library(freertos_config INTERFACE)\n"
+            "  target_include_directories(freertos_config SYSTEM\n"
+            "    INTERFACE\n"
+            "      include) # The config file directory\n"
+            "  target_compile_definitions(freertos_config\n"
+            "    PUBLIC\n"
+            "    projCOVERAGE_TEST=0)\n")
+    else()
+        message(WARNING " Using deprecated 'FREERTOS_CONFIG_FILE_DIRECTORY' - please update your project CMakeLists.txt file:\n"
+            "  add_library(freertos_config INTERFACE)\n"
+            "  target_include_directories(freertos_config SYSTEM\n"
+            "    INTERFACE\n"
+            "      include) # The config file directory\n"
+            "  target_compile_definitions(freertos_config\n"
+            "    PUBLIC\n"
+            "    projCOVERAGE_TEST=0)\n")
+        # Currently will add this in here.
+        add_library(freertos_config INTERFACE)
```

Comment:
I would much prefer that we leave this up to the caller to handle and just deprecate FREERTOS_CONFIG_FILE_DIRECTORY.

Then the caller can add their relevant FreeRTOSConfig.h file and directory to the target in whichever way they see fit.